### ISSUE TITLE
launcher: URL not valid when launching workflows without being logged in

### ISSUE THREAD

**Issue by mdonadoni** [reanahub/reana-server#482]
How to reproduce:
1. Make sure you are not logged in to [reana.cern.ch](https://reana.cern.ch)
2. Click on the `Launch on REANA` badge in [reana-demo-atlas-recast](https://github.com/reanahub/reana-demo-atlas-recast)
3. Login using the CERN SSO
4. Click on the `Launch` button
5. The workflow is not launched: `Field 'url': Not a valid URL.`

**Comment by tiborsimko** [comment]
When working on this issue, please check also the encoding of other arguments. E.g. `name=foo%20bar` is silently accepted, and then the space is dropped, so the workflow run is created, it'll be named `foobar` without space. Whilst one would expect the space not to be dropped, because via CLI it is perfectly possible to do:

```console
$ reana-client run -w 'foo bar'
```

and the worfklow run will be happily named `foo bar.1` with space included.

**Comment by VMois** [comment]
I think [this line](https://github.com/reanahub/reana-ui/blob/61ae514d0ae6aa75188f6e4c02e1ad2f8a7081e3/reana-ui/src/client.js#L20) of `reana-ui` is responsible for encoding return URL. And because the URL was already encoded it will do it twice.

**Comment by VMois** [comment]
There are a few solutions:

- do not encode URLs in our badges
- remove `stringify` from `reana-ui`, @mdonadoni is it possible without breaking it?

**Comment by tiborsimko** [comment]
> do not encode URLs in our badges

That is not a solution in my eyes. As we have discussed yesterday, `/` is a reserved character, and its replacement is perfectly valid. For example, try to search in Google for `a/b` and you will see that it leads to a URL having `https://www.google.com/search?q=a%2Fb` with the `/` sign being encoded.  So if you create a link to that Google search, you would naturally encode the `/` signs.

Our current badge URLs like `https://reana.cern.ch/launch?url=https%3A%2F%2Fgithub.com%2Freanahub%2Freana-demo-atlas-recast&name=reana-demo-atlas-recast` are perfectly valid RFC 2396 URIs, so we should support them.

IMO the frontend app should properly read and decode parameters, which will "unwrap" `url` values into the usual `https://...`, textual form, and then when it calls the backend it can pass the proper values.

This should work well for both style of badge URLs, encoded and not encoded. Please see a naive example in Python:

```python
In [13]: url1 = 'https://reana.cern.ch/launch?url=https%3A%2F%2Fgithub.com%2Freanahub%2Freana-demo-atlas-recast&name=reana-demo-atlas-recast'

In [14]: url2 = 'https://reana.cern.ch/launch?url=https://github.com/reanahub/reana-demo-atlas-recast&name=reana-demo-atlas-recast'

In [15]: urllib.parse.parse_qs(url1)
Out[15]:
{'https://reana.cern.ch/launch?url': ['https://github.com/reanahub/reana-demo-atlas-recast'],
 'name': ['reana-demo-atlas-recast']}

In [16]: urllib.parse.parse_qs(url2)
Out[16]:
{'https://reana.cern.ch/launch?url': ['https://github.com/reanahub/reana-demo-atlas-recast'],
 'name': ['reana-demo-atlas-recast']}
```

(disregard the overall simple logic; just check the URL expansion)

**Comment by VMois** [comment]
Yes, it is not the most correct but, probably, the fastest solution. Unless we will wait to deploy the new SSO. As we discussed with @mdonadoni, the issue is probably with how OAuth is decoding URL.

The below URL is encoded once.

```
https://reana.cern.ch/launch?url=https%3A%2F%2Fgithub.com%2Freanahub%2Freana-demo-atlas-recast&name=reana-demo-atlas-recast
```

But what happens before redirecting the user to the SSO login page is that the URL gets encoded a second time, and is passed as `next` query parameter to SSO. UI will use this to redirect after login:

```python
https://reana.cern.ch/oauth/login/cern?next=%2Flaunch%3Furl%3Dhttps%253A%252F%252Fgithub.com%252Freanahub%252Freana-demo-atlas-recast%26name%3Dreana-demo-atlas-recast
```

This is shouldn't be a problem, but for some reason when login is complete, SSO returns back a partially undecoded URL:

```json
{
  "app": "cern",
  "next": "/launch?url=https%253A%252F%252Fgithub.com%252Freanahub%252Freana-demo-atlas-recast&name=reana-demo-atlas-recast",
  "sid": "..."
}
``` 

You can see that `/launch?...` part is decoded, but the `https%253A...` is not at all. 

Thanks @mdonadoni for the info.

**Comment by VMois** [comment]
(1) I think, it is a dirty fix, and we probably need to do it on the UI side.

(2) Looks like it behaves better, but the new SSO is not even enconding URL so it works, but not technically valid. We will need some changes to UI anyway to make SSO redirect correctly. More info @mdonadoni
